Add '--clang' flag to runner.py

rapidsna opened the pull request:
### Pull Request Description

'--clang' provides a way to build projects with a custom clang binary.
With '--clang-source-path', the 'run' script builds clang binary from
the clang source in the path and provides this binary path to 'runner.py'
using the '--clang' flag.

### Acceptance Criteria

To be accepted into the Swift source compatibility test suite, a project must:

- [ ] be an *Xcode* or *swift package manager* project
- [ ] support building on either Linux or macOS
- [ ] target Linux, macOS, or iOS/tvOS/watchOS device
- [ ] be contained in a publicly accessible git repository
- [ ] maintain a project branch that builds against Swift 4.0 and passes any unit tests
- [ ] have maintainers who will commit to resolve issues in a timely manner
- [ ] be compatible with the latest GM/Beta versions of *Xcode* and *swiftpm*
- [ ] add value not already included in the suite
- [ ] be licensed with one of the following permissive licenses:
	* BSD
	* MIT
	* Apache License, version 2.0
	* Eclipse Public License
	* Mozilla Public License (MPL) 1.1
	* MPL 2.0
	* CDDL
- [ ] pass `./project_precommit_check` script run

Ensure project meets all listed requirements before submitting a pull request.

rapidsna commented:
@swift-ci test

rapidsna commented:
@swift-ci test

ahoppen commented on clang_installer.py:
```diff
@@ -0,0 +1,84 @@
+        for clang_path in self.clang_paths:
+            clang_orig_path = clang_path + self.orig_postfix()
+            if not os.path.isfile(clang_orig_path):
+                continue
```
AFAICT this should never happen, right? We expect that we always moved an original clang executable to the safe location, right? Wouldn’t an assertion failure be better instead of silently skipping over?

ahoppen commented on clang_installer.py:
```diff
@@ -0,0 +1,84 @@
+                if os.path.islink(clang_path):
+                    common.check_execute(['rm', clang_path])
+                else:
+                    continue
```
Isn’t this an error case as well if we are running `clear` without having a symlink at `clang_path`? So wouldn’t an assertion failure be better?

ahoppen commented on run:
```diff
@@ -37,11 +37,19 @@ def main():
+        if args.clang_source_path:
+            clangc = build_clang(workspace, args.clang_source_path)
+        if clangc:
+            clang_flag = ['--clangc', clangc]
```
I think `clang_flags` (plural) would be better as this could contain multiple flags. Same in `execute_runner` and `execute_build_incremental` below.

rapidsna replied:
Thanks! I renamed it to `clang_flags`.

ahoppen replied:
I thought we could just use
```
common-check_execute([ninja_path])
```
and avoid looking up `ninja` through `xcrun` again, but of course that doesn’t set the SDK environment variables etc.

ahoppen commented on run:
```diff
@@ -74,6 +82,15 @@ def parse_args():
+    clang_arguments = parser.add_mutually_exclusive_group()
+    clang_arguments.add_argument('--clang-source-path',
+                                 metavar="PATH",
+                                 help='path to llvm-project to build',
```
Suggestion for improved message:
```suggestion
                                 help='If specified, builds a new clang executable from the given llvm-project directory and uses it to build the suite',
```

rapidsna replied:
Updated!

ahoppen commented on run:
```diff
@@ -74,6 +82,15 @@ def parse_args():
+                                 type=os.path.abspath)
+    clang_arguments.add_argument('--clangc',
+                                 metavar="PATH",
+                                 help='clang executable',
```
Suggestion for improved message:
```suggestion
                                 help='If specified uses the given clang executable to build the suite',
```

ahoppen replied:
Just a heads-up that I suggested an improved message here. If you prefer your one, just comment.

rapidsna replied:
I updated it to be "clang executable to build Xcode projects".

ahoppen reviewed (commented):
The whole approach of swapping a `clang` executable into the toolchain seems awfully fragile to me (see the amount of comments I left in that file). Is there a reason why you can’t just specify the `clang` executable by setting the `CC` environment variable similar to how we set `SWIFT_EXEC`?

I also added some comments regarding code style inline.

rapidsna commented:
Thanks @ahoppen for reviewing my PR!

> The whole approach of swapping a `clang` executable into the toolchain seems awfully fragile to me (see the amount of comments I left in that file). Is there a reason why you can’t just specify the `clang` executable by setting the `CC` environment variable similar to how we set `SWIFT_EXEC`?

The reason why we couldn't just specify a `clang` executable path is because that doesn't work with Xcode project build. xcodebuild always finds the clang binary from a toolchain location so we need to have a whole toolchain or swap the clang executables in the toolchain path. Some users may want to test the build with their custom clang and I don't expect users should generate a whole xcode toolchain so being able to specify their custom clang executable will be very useful.

> I also added some comments regarding code style inline.

Thanks!

rapidsna commented:
@swift-ci test

rapidsna commented:
@swift-ci test

rapidsna commented:
Thanks @ahoppen I addressed the style comments, and as you pointed out it now adds `CC=clangc` to Xcode flags instead of binary replacement. I removed the code that does it.

ahoppen commented on run:
```diff
@@ -270,6 +292,28 @@ def build_swift_toolchain(workspace, args):
+                '-DLLVM_TARGETS_TO_BUILD=X86;AArch64;ARM',
+                source_path]
+            common.check_execute(cmake_command)
+            common.check_execute(['xcrun', 'ninja'])
```
I suppose you could just use `ninja_path` here again, right? There’s no need to look it up again through `xcrun`. Would also make it more robust IMHO if the user `xcode-select`ed a different version of Xcode between the two runs.

rapidsna replied:
`common.check_execute(['xcrun', 'ninja'])` is to build the project with ninja. I'll add some comments.

ahoppen commented on run:
```diff
@@ -270,6 +292,28 @@ def build_swift_toolchain(workspace, args):
+            common.check_execute(cmake_command)
+            common.check_execute(['xcrun', 'ninja'])
+        except:
+            raise
```
Why is the `try catch` block needed if you are just raising the exception?

rapidsna replied:
Removed it. Thanks!

ahoppen commented on run:
```diff
@@ -193,10 +212,12 @@ def execute_runner(workspace, args):
 
 
-def execute_build_incremental(workspace, args):
+def execute_build_incremental(workspace, args, clang_flags):
```
Same as for `execute_runner`

ahoppen commented on run:
```diff
@@ -161,7 +180,7 @@ def get_sandbox_profile_flags_test():
 
 
-def execute_runner(workspace, args):
+def execute_runner(workspace, args, clang_flags):
```
Sorry for revising my last review, but I just realized `clang_flags` isn’t really restricted to `clang`-specific flags, nor does it pass these flags to `clang`. So, probably `additional_args`  would be a better name.

rapidsna replied:
Yea, for the caller perspective it is clang flags, for callees additional_args makes sense.

rapidsna replied:
I use `additional_runner_args` since unlike `args`, the arguments are passed to `runner` command.

ahoppen replied:
The setting makes sense then. I was mostly curious how those relate and your explanation makes total sense. Thanks!

ahoppen commented on run:
```diff
@@ -270,6 +292,28 @@ def build_swift_toolchain(workspace, args):
+                '-DLLVM_ENABLE_PROJECTS=clang;llvm',
+                '-DCMAKE_BUILD_TYPE=Release',
+                '-DCLANG_APPLE_BUILD_VERSION_STRING=13000000',
+                '-DLLVM_TARGETS_TO_BUILD=X86;AArch64;ARM',
```
No really blocking the PR but how did you come up with building for these architectures? IIUC we are building Swift for the host platform + `macosx-arm64` (based on https://github.com/apple/swift/blob/5e9ff1596061cfbde9c0ca0bc7497570914da6c3/utils/build-presets.ini#L2569).

I’m not entirely sure how this relates to the architectures you specified.

rapidsna replied:
Specifying possible compiler targets is different from setting cross-compilation hosts, the former is about what targets the compiler can compile for and the latter is about what host architecture the compiler executable itself runs on.

Hence, `DLLVM_TARGETS_TO_BUILD` is more related to `llvm-targets-to-build=X86;ARM;AArch64;PowerPC` in https://github.com/apple/swift/blob/5e9ff1596061cfbde9c0ca0bc7497570914da6c3/utils/build-presets.ini#L594

We just don't add `PowerPC` as we don't have adopters run on it.

ahoppen commented on run:
```diff
@@ -270,6 +292,28 @@ def build_swift_toolchain(workspace, args):
+                '-DCMAKE_MAKE_PROGRAM={}'.format(ninja_path),
+                '-DLLVM_ENABLE_PROJECTS=clang;llvm',
+                '-DCMAKE_BUILD_TYPE=Release',
+                '-DCLANG_APPLE_BUILD_VERSION_STRING=13000000',
```
I suppose you are hardcoding this version number because it’s the current clang version and it’s bound to get out of date at some point. I’d prefer something completely non-sensical like `99999999` or to remove it, if possible. Otherwise we’ll wonder in a couple of years why this needs to be `13000000`.

rapidsna replied:
That specific string is necessary for building an internal project, but I removed it from this PR.

ahoppen commented on run:
```diff
@@ -270,6 +292,28 @@ def build_swift_toolchain(workspace, args):
+                'xcrun', 'cmake', '-G', 'Ninja',
+                '-DCMAKE_MAKE_PROGRAM={}'.format(ninja_path),
+                '-DLLVM_ENABLE_PROJECTS=clang;llvm',
+                '-DCMAKE_BUILD_TYPE=Release',
```
Swift decides whether to build a `Debug`, `Debug+Asserts`, `Release` or `Release+Asserts` version of itself depending on whether `--debug` and `--assertions` are specified.
 https://github.com/apple/swift-source-compat-suite/blob/43abe0b6035e28088e9bd049869f359467e1d2d4/run#L226-L233
Would it make sense to do the same here? 
If yes, the argument descriptions of the `--debug` and `--assertions` flags need to be updated.
If not, I think it would make sense to add a note to the `clang-source-path`, telling that `clang` will always be build in `Release` mode without assertions.

rapidsna replied:
I updated so that the build type is determined by the debug and assertions flags. I've also updated `--debug` and `--assertions` description that it may also apply to clang.

ahoppen commented on project_future.py:
```diff
@@ -584,6 +584,9 @@ def add_arguments(parser):
     parser.add_argument("--report-time-path",
                         help='export time for building each xcode build target to the specified json file',
                         type=os.path.abspath)
+    parser.add_argument("--clangc",
```
Does it make sense to name this argument `clangc` instead of `clang`? With the current naming, I expect I should pass in a `clangc` executable (which doesn’t exist AFAICT). 

Note that for Swift we actually have two different symlinks to the compiler: `swift` which can also do things like `swift build` and `swiftc` which is the compiler itself, so there’s a difference.

rapidsna replied:
I updated to use `clang` instead of `clangc`.

ahoppen reviewed (commented):
I think this is **much** cleaner with the `CC` environment variable now. Thanks!

I’ve added some more comments inline.

ahoppen reviewed (approved):
Thanks for being patient and addressing all my comments. Looks good to me now 👍
